runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/GrpcDirectGetWorkStream.java

```diff
@@ -267,25 +277,100 @@ protected void startThrottleTimer() {
   }
 
   @Override
-  public void adjustBudget(long itemsDelta, long bytesDelta) {
-    GetWorkBudget adjustment =
-        nextBudgetAdjustment
-            // Get the current value, and reset the nextBudgetAdjustment. This will be set again
-            // when adjustBudget is called.
-            .getAndUpdate(unused -> GetWorkBudget.noBudget())
-            .apply(itemsDelta, bytesDelta);
-    sendRequestExtension(adjustment);
+  public void setBudget(long newItems, long newBytes) {
+    GetWorkBudget currentMaxGetWorkBudget =
+        maxGetWorkBudget.updateAndGet(
+            ignored -> GetWorkBudget.builder().setItems(newItems).setBytes(newBytes).build());
+    GetWorkBudget extension = budgetTracker.computeBudgetExtension(currentMaxGetWorkBudget);
+    maybeSendRequestExtension(extension);
   }
 
-  @Override
-  public GetWorkBudget remainingBudget() {
-    // Snapshot the current budgets.
-    GetWorkBudget currentPendingResponseBudget = pendingResponseBudget.get();
-    GetWorkBudget currentNextBudgetAdjustment = nextBudgetAdjustment.get();
-    GetWorkBudget currentInflightBudget = inFlightBudget.get();
-
-    return currentPendingResponseBudget
-        .apply(currentNextBudgetAdjustment)
-        .apply(currentInflightBudget);
+  /**
+   * Tracks sent and received GetWorkBudget and uses this information to generate request
+   * extensions.
+   */
+  @AutoValue
+  abstract static class GetWorkBudgetTracker {
+
+    private static GetWorkBudgetTracker create() {
+      return new AutoValue_GrpcDirectGetWorkStream_GetWorkBudgetTracker(
+          new AtomicLong(), new AtomicLong(), new AtomicLong(), new AtomicLong());
+    }
+
+    abstract AtomicLong itemsRequested();
```

Comment:
how about just using synchronized instead of lots of separate atomics? Multiple atomic ops might be worse performance anyway and it means we might have weird races where they are inconsistently updated.

Reply:
Done also added tests for this class

Reply:
done